File: Lee/LeetCode/643. Maximum Average Subarray I/643.cpp
#include <algorithm>
#include <iterator>
#include <numeric>
#include <ranges>
#include <tuple>
#include <vector>

class Solution
{
public:
    double findMaxAverage(std::vector<int>& nums, int k)
    {
        auto state = std::reduce(std::cbegin(nums), std::next(std::cbegin(nums), k));
        auto maximum = state;

        for (auto p : std::views::zip(nums, std::views::drop(nums, k)))
        {
            state = state - std::get<0>(p) + std::get<1>(p);
            maximum = std::max(maximum, state);
        }

        return static_cast<double>(maximum) / static_cast<double>(k);
    }
};

File: Lee/LeetCode/643. Maximum Average Subarray I/643.rs
use std::{cmp::max, iter::zip};

impl Solution {
    pub fn find_max_average(nums: Vec<i32>, k: i32) -> f64 {
        let n = usize::try_from(k).unwrap();

        let mut state: i32 = nums.iter().take(n).copied().sum();
        let mut maximum = state;

        for (left, right) in zip(nums.iter(), nums.iter().skip(n)) {
            state = state - left + right;
            maximum = max(maximum, state);
        }

        f64::from(maximum) / f64::from(k)
    }
}
